---
layout: default
title: Loading and filtering
---

<!-- stuff to make Rmarkdown do what we want:  -->
```{r setup, include=FALSE}
suppressMessages(library(rmarkdown))
knitr::opts_chunk$set(echo = TRUE,
  fig.width=9,
  fig.height=6,
  cache=FALSE, # only use TRUE if you're really sure. Also check xfun::cache_rds
  collapse=TRUE,  # put misc. outputs from one chunk in one block
  tidy=TRUE, # tidies up the R code
  tidy.opts=list(arrow=TRUE, 
                 indent=2,
                 width.cutoff = 60))
```

Today we will be looking at Rhabdomyosarcoma data generated by Jeff deMartino. The publication (in press) is:

DeMartino, Jeff, Michael T. Meister, Lindy Visser, Mariël Brok, Marian J. A. Groot Koerkamp, Laura S. Hiemcke-Jiwa, Terezinha de Souza, et al. ‘Single-Cell Transcriptomics Reveals Immune Suppression and Cell States Predictive of Patient Outcomes in Rhabdomyosarcoma’. bioRxiv, 18 July 2022. https://doi.org/10.1101/2022.07.15.497944.

<!-- load complete state from previous lesson ; not the first time-->

# Loading and filtering data
## Preliminary stuff

```{r libraries, warning=FALSE, message=FALSE}
# libraries
library(Seurat)
library(Matrix)
library(dplyr)
library(ggplot2)
library(patchwork)
library(clusterProfiler) # enrichment analysis
library(org.Hs.eg.db) # gene annotation databases
library(SCutils) # our R package with various convenience data and functions.
library(SingleR) # automatic celltype identification
library(infercnv) # finding CNV's in scRNAseq data
library(clustree) # visualize how clusters split when increasing resolution
library(ggraph) # needed for clustree
```

```{r configvars}

data(refdata_cellranger_GRCh38_3.0.0) # reference data

CFG <- list() # global

CFG$data_dir <- "/opt/course/day2" # where to ready data from
CFG$output_dir <- "."              # where to write data to

CFG$random_seed <- 2033

```

```{r overrule-dirs, include=FALSE}
## when developing, data_dir and output_dir may be different; overrule them here:
if ( Sys.getenv("dev") == "TRUE" ) {
  CFG$data_dir <-  "~/data/course/day2"
  CFG$output_dir <- "~/tmp/day2"
}
```

## Loading data

The 384-well plate-based platform that was used for these samples
includes empty wells and control features. For convenience, we provide
you with a ready-made Seurat object that omits these empty wells and
control features.

```{r readdata}
## Load it:
file <- paste0(CFG$data_dir,"/srat-raw.rds")
srat <- readRDS(file)

```

## Filter cells

The next step, a QC-step really, is to remove low quality cells, as we did yesterday.

## Minimum transcripts

```{r filtercells}

## show the distribution of transcipts, by plate:
v <- VlnPlot(srat, "nCount_RNA", group.by = 'plate_id')

v

## same, but using a logarithmic scale:
v + scale_y_continuous(trans='log2')

## following might be reasonable:

CFG$min_txpts = 1000
CFG$max_txpts = Inf  # meaning Infinity

```

## Max transcripts

Some of the plates have cells that might seem outliers, but they are
different per plate. You could get rid of cells with e.g. >=
50,000 transcripts[^percentile].

[^percentile]: An alternative would be to throw out the cells in the top 0.5% of transcript counts.

<!-- for code throwing out 0.5% top percentile per plate, see prior to 21-Jan-2022 07:19:48 -->

**Exercise 1:** Can we omit a maximum number of transcripts in this dataset?

## Mitochondrial percentage

```{r mitochondria}

v <- VlnPlot(srat, "percent_mito", group.by = 'plate_id')

v

CFG$max_pctmito = 60

```

## Final selection

```{r finalselect}

## show our choices:
FeatureScatter(srat, feature1='nCount_RNA', 
  feature2='percent_mito', pt.size=2)  + 
  geom_hline(yintercept = CFG$max_pctmito, linetype=2) +
  geom_vline(xintercept = c(CFG$min_txpts, CFG$max_txpts),linetype=2) + 
  scale_x_continuous(trans='log2')

## how much do we have currently:
dim(srat)

## what will we lose:
dim(subset(srat, nCount_RNA < CFG$min_txpts))
dim(subset(srat, percent_mito > CFG$max_pctmito))

## and do the subsetting
srat <- subset(srat,
               subset =
                 nCount_RNA >= CFG$min_txpts
               & percent_mito <= CFG$max_pctmito ) 

## how much do we have now:
dim(srat)

# maybe save
file  <- paste0(CFG$output_dir,'/srat-cellsfiltered.rds')
saveRDS(file=file, srat)

```



<!-- lastly, save the complete sesssion for the next time -->
```{r savesession, include=FALSE} 
## save complete state for next lesson
## save complete state for next lesson
## save.image("session.rda")
if(Sys.getenv("nodump")!="TRUE")
    save(file="session.rda", list=c("srat", ls(pat="col|CFG")))
```
