### autogenerated from R -e 'knitr::purl(01-loadingdata.Rmd)'
### autogenerated from R -e 'knitr::purl(01-loadingdata.Rmd)'
### autogenerated from R -e 'knitr::purl(01-loadingdata.Rmd)'
## ----setup, include=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
suppressMessages(library(rmarkdown))
knitr::opts_chunk$set(echo = TRUE,
  fig.width=9,
  fig.height=6,
  cache=FALSE, # only use TRUE if you're really sure. Also check xfun::cache_rds
  collapse=TRUE,  # put misc. outputs from one chunk in one block
  tidy=TRUE, # tidies up the R code
  tidy.opts=list(arrow=TRUE,
                 indent=2,
                 width.cutoff = 60))


## ----libraries, warning=FALSE, message=FALSE------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

library(Seurat) # our main 'environment' 
library(Matrix) # needed for working with sparse matrices
library(dplyr)  # general data manipulation
library(ggplot2) # plotting functions
library(patchwork) # composing multi-panel plots
library(org.Hs.eg.db) # genome information
library(SCutils) # our R package with various convenience data and functions.


## ----configvars-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

CFG <- list() # global

CFG$data_dir <- "/opt/course/day1" # where to read data from
CFG$output_dir <- "."              # where to write data to

# explained later:
CFG$ndims <- 25
CFG$random_seed <- 2033



## ----overrule-dirs, include=FALSE-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
## when developing, data_dir and output_dir may be different; overrule them here:
if ( Sys.getenv("dev") == "TRUE" ) {
  CFG$data_dir <-  "~/data/course/day1"
  CFG$output_dir <- "~/tmp/day1"
}


## ----params---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
# Cutoffs used during quality control for filtering genes (explained later)
CFG$min_txpts <- 5000 
CFG$max_txpts <- 60000
CFG$max_pctmito <- 25

# Graphical parameters

CFG$gradient.colors = viridis::viridis(101, direction= -1)

## NOTE: many colors change if the number of clusters change, so
## they are not fixed here.



## ----loading--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
lib <- "TX010" # name of the library, also needed later on.
dir <- paste0(CFG$data_dir, "/", lib) # what is actually on disk
counts <- Read10X(data.dir=dir) # this is of type 'dgCMatrix', a sparse matrix


## ----txpts-per-cell-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

cnts_per_cell <- Matrix::colSums(counts) # note use of Matrix::colSums

summary(cnts_per_cell)

df <- data.frame(counts=cnts_per_cell)

ggplot(df, aes(x=counts)) +
  geom_density() +
  geom_rug() +
  scale_x_continuous(trans='log2')



## ----change-cell-names----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  
  cellnames <- colnames(counts)
  colnames(counts) <- paste0(lib, "_", cellnames)



## ----metadata-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

# get the mitochondrial genes based on their name:
mitos <- grep("^mt-", rownames(counts), value=TRUE, ignore.case=TRUE) 

## show them:
mitos

percent_mito <- 100 * Matrix::colSums(counts[mitos, ]) / cnts_per_cell # note the use of Matrix::colSums again

## get the non-mito's for easy selection of 'the rest of the genes'
nuclear <- setdiff(rownames(counts), mitos)

log2_transcript_counts <- log2(1+Matrix::colSums(counts[nuclear , ]))

log2_feature_counts <- log2(1+Matrix::colSums(counts[nuclear, ] > 0))

## set up a data.frame:
meta <- data.frame(percent_mito = percent_mito,
  log2_counts = log2_transcript_counts,
  log2_features = log2_feature_counts,
  lib=rep(lib, ncol(counts)))

## let's show how features ( = genes ) relate to number of transcripts
ggplot(meta, aes(x=log2_counts, y=log2_features)) + geom_point(color='blue')



## ----create-seurat--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

counts <- counts[ nuclear , ]

srat <- Seurat::CreateSeuratObject(counts=counts,
    project = 'intestinal_organoids',
    meta = meta)

## Later this we won't use the counts object anymore, so 
## we should get rid of it and cleanup
rm(counts)
gc()



## ----str-obj--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
srat # just an overview

str(srat)
 


## ----head-metadata--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

head(srat@meta.data) ## short-hand: head(srat)



## ----savesession, include=FALSE-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

## save complete state for next lesson
if(Sys.getenv("nodump")!="TRUE")
  save(file="session.rda", list=c("srat", ls(pat="col|CFG")))

