--- 
layout: default
title: Hemoglobin removal
--- 


<!-- stuff to make Rmarkdown do what we want:  -->
```{r setup, include=FALSE}
suppressMessages(library(rmarkdown))
knitr::opts_chunk$set(echo = TRUE,
  fig.width=9,
  fig.height=6,
  cache=FALSE, # only use TRUE if you're really sure. Also check xfun::cache_rds
  collapse=TRUE,  # put misc. outputs from one chunk in one block
  tidy=TRUE, # tidies up the R code
  tidy.opts=list(arrow=TRUE, 
                 indent=2,
                 width.cutoff = 60))
```

<!-- load complete state from previous lesson -->
```{r loadsession, include=FALSE} 
source('../libs.R')
load("../09-loadfilter/session.rda")
```

# Hemoglobin removal 

When plotting the number of genes versus the number of transcripts per plate,
something unusual is going on. 

```{r features-vs-counts}

plates <- unique(srat@meta.data$plate_id)

## for convenience we will plot everything in a loop:
for (plate in plates) { 
  df <- subset( srat@meta.data, plate_id==plate)
  p =  ggplot(df,
         aes(x=nCount_RNA, y=nFeature_RNA, color=plate_id)) + geom_point() + 
         labs(title=plate)
  
  ## ggplot2 plot objects won't plot automaticlly inside a loop.
  ## We therefore have explicitly call it with show():
  
  show(p)

  ## and/or logarithmic:

  p + scale_x_continuous(trans='log2') + 
     scale_y_continuous(trans='log2')
}

```

In a few of these plots you can clear see cells lying below the curve. 

**Exercise 1:** What is different about these cells (in terms of numbers of transcripts and genes), and how would you interpret it?

<!-- These 'off-diagonal' cells express the same number of transcripts with
fewer genes, or converse, they express more transcripts given the same number of 
genes. IOW., they are more specialized. -->

We have to check what is going on here. 

## Selecting cells

A useful way to explore groups of cells is to use the `CellSelector`
function to interactively pick cells from a plot[^cellselector].  Having
them, you can investigate them using e.g. `FindMarkers`.  It works as
follows:

[^cellselector]: even though we use `CellSelector` on a plot that was not produced by Seurat, it still works. The reason is that Seurat uses ggplot2 throughout, which comes with this capability, provided the data.frame used for the plot has proper rownames (in our case are simply the cell IDs).

```{r cellselector, eval=FALSE}

selected <- CellSelector(plot = p)  # this launches the cell picker. 
## To complete it, click the Done button. 

## The variable `selected` now contains the names of the picked cells
show(selected)

## If your override their Identity you can use
## FindMarkers to find differentially expressed genes:
Idents(srat, cells = selected ) <- "handpicked"

markers <- FindMarkers(srat, ident.1 = "handpicked",  only.pos = TRUE)

# show results, selecting on p_value and log fold change:
subset(markers, p_val_adj < 1e-2 &  avg_log2FC > 2)

```

## Finding the contamination

These 'off-diagonal' cells express the same number of transcripts with
fewer genes, i.e. they are more specialized. A typical example of such
cells are erythrocytes containing predominantly hemoglobin mRNA's.
The cells we see here may be be proper cells that are contaminated 
with ambient hemoglobin mRNA from erythrocytes.

Let's show this and filter them out. For this, we can again use the
ready-made lists of genes that are available from the `SCutils` package.
One of the lists is `genelists$hemo`, but you can also use the function
`SCutils::lookup.hemogenes()`

We will sum the transcripts of all the hemoglobin-related genes, but
not all of them are present in our Seurat object. You therefore have to
limit the list to those that are present using the `intersect` function.

```{r show-hemo-contam}

library(SCutils)
data(refdata_cellranger_GRCh38_3.0.0)

hb_genes <- genelists$hemo 
## alternatively:
# hb_genes <- lookup.hemogenes()

hb_genes <- intersect(hb_genes, rownames(srat))

hb_counts <- Matrix::colSums(srat@assays$RNA@counts[ hb_genes, ])

srat <- AddMetaData(srat,
                    col.name = "log2hb_genes",
                    metadata=log2(1+hb_counts))

srat <- AddMetaData(srat,
                    col.name = "pct_hemo",
                    metadata=100*hb_counts/srat@meta.data$nCount_RNA )

## now show the previous plot along side a plot of the hemo content:
for (plate in plates) {

  df <- srat@meta.data[  srat@meta.data$plate_id == plate, ]
  
  p_ngenes <- ggplot(df, aes(x=log2_counts, y=log2_features, color= pct_hemo)) + 
    geom_point(size=1, alpha=0.5) +
    scale_color_gradient(low="blue", high="red", limits=c(0,5), oob=scales::squish) +
    labs(title=plate)

  p_pcthemo <- ggplot(df, aes(x=log2_counts, y=pct_hemo, color= log2_features)) + 
     geom_point(size=1, alpha=0.5) +
     scale_color_gradient(low="blue", high="red", limits=c(2,16), oob=scales::squish)

## again using show() because we're inside a loop
show(p_ngenes | p_pcthemo)

}
```

## Do the removal

Having seen the data we could decide to throw out any cells with more than 5 percent hemoglobin genes.

```{r hemo-removal}
## set a maximum to get rid of cells, and do the selection
CFG$max_pct_hemo <- 5                   #

dim(srat)
dim(subset(srat, pct_hemo > CFG$max_pct_hemo))

srat <- subset(srat, pct_hemo <= CFG$max_pct_hemo)

## maybe save
file <- paste0(CFG$output_dir, "/srat-cellsfiltered.rds")
saveRDS(file=file, srat)

```


<!-- lastly, save the complete sesssion for the next time -->
```{r savesession, include=FALSE} 
## save complete state for next lesson
## save complete state for next lesson
## save.image("session.rda")
if(Sys.getenv("nodump")!="TRUE")
    save(file="session.rda", list=c("srat", ls(pat="col|CFG")))
```
